_source/_posts/2024-01-24-spring-boot-authorization.md

```diff
@@ -0,0 +1,485 @@
+
+For learning purposes, let's assume you have built a Spring Boot menu API that must be secured, so only authorized users can perform requests to its endpoints. Now you are going to implement authorization for the API with OAuth2 2.0 and Auth0. Start by doing a checkout of the API repository, which already implements basic request handling:
+
+```shell
+git clone https://github.com/indiepopart/spring-menu-api.git
+```
+
+The repository contains two project folders, `start` and `demo`. The bare bones menu API is a Gradle project in the `start` folder, open it with your favorite IDE. If you would rather skip the step-by-step security configuration and just run the final `demo` project, follow the instructions in the [README](https://github.com/indiepopart/spring-menu-api).
+
+Sign up at [Auth0](https://auth0.com/signup) and install the [Auth0 CLI](https://github.com/auth0/auth0-cli). Then in the command line run:
+
+```shell
+auth0 login
+```
+
+The command output will display a device confirmation code and open a browser session to activate the device.
+
+You don't need to create a client application for your API if not using opaque tokens. But you must register the API within your tenant, you can do it using Auth0 CLI:
+
+```shell
+auth0 apis create \
+  --name "Menu API" \
+  --identifier https://menu-api.okta.com \
+  --scopes "create:items,update:items,delete:items" \
+  --offline-access=false
+```
+
+The scopes `create:items`, `update:items`, `delete:items` will be required ahead in the tutorial. Add the `okta-spring-boot-starter` dependency:
+
+```groovy
+// build.gradle
+dependencies {
+    ...
+    implementation 'com.okta.spring:okta-spring-boot-starter:3.0.6'
+    ...
+}
+```
+As the `menu-api` must be configured as an OAuth2 resource server, add the following properties:
+
+```properties
+# application.properties
+okta.oauth2.issuer=https://<your-auth0-domain>/
+okta.oauth2.audience=https://menu-api.okta.com
+```
+
+```shell
+auth0 tenants list
+```
+
+Run the API with:
+
+```shell
+./gradlew bootRun
+```
+
+Test the API authorization with curl:
+
+```shell
+curl localhost:8080/api/menu/items
+```
+You will get HTTP response code `401` because the request requires bearer authentication. Using Auth0 CLI, get an access token:
+
+```shell
+auth0 test token -a https://<your-auth0-domain>/api/v2/ -s openid
+```
+Select any available client when prompted. You also will be prompted to open a browser window and log in with a user credential. You can sign up as a new user using an email and password or using the Google social login.
+
+With curl, send a request to the API server using a bearer access token:
+
+```shell
+ACCESS_TOKEN=<auth0-access-token>
```

Comment:
both commands can be merged in one block

Reply:
Updated. Combined in one line.